### PR TITLE
Various code review changes

### PR DESCRIPTION
**Pull request by nicowilliams**
These changes come from Christos Zoulas.

### CODE REVIEW

File: lib/krb5/send_to_kdc.c [hunk 1]
```diff
@@ -1244,12 +1244,16 @@ krb5_sendto_context(krb5_context context,
     }
 
     _krb5_debug(context, 1,
-		"krb5_sendto_context %s done: %d hosts %lu packets %lu wc: %ld.%06ld nr: %ld.%06ld kh: %ld.%06ld tid: %08x",
-		realm, ret,
+		"%s %s done: %d hosts %lu packets %lu:"
```
Comment (elric1) [inline lib/krb5/send_to_kdc.c]: You've added a colon here after %lu.

File: lib/kadm5/ad.c [hunk 2]
```diff
@@ -1286,45 +1286,44 @@ kadm5_ad_randkey_principal(void *server_handle,
 	return ret;
     }
 
-    krb5_data_zero (&result_code_string);
-    krb5_data_zero (&result_string);
-
-    ret = krb5_set_password_using_ccache (context->context,
```
Comment (elric1) [inline lib/kadm5/ad.c]: Lots of unrelated whitespace changes in this commit.

File: lib/kadm5/ad.c [hunk 3]
```diff
@@ -1286,45 +1286,44 @@ kadm5_ad_randkey_principal(void *server_handle,
...
+    ret = krb5_string_to_key(context->context,
+                             ENCTYPE_ARCFOUR_HMAC_MD5,
+                             password,
+                             principal,
+                             &(*keys)[0]);
+    if (ret) {
+        free(*keys);
+        *keys = NULL;
+        *n_keys = 0;
+        goto out;
```
Comment (elric1) [inline lib/kadm5/ad.c]: goto out; is unnecessary?

File: kpasswd/kpasswdd.c [hunk 4]
```diff
@@ -646,6 +646,32 @@ process (krb5_keytab keytab,
...
+ * XXX this code relies on getsockname() returning a valid local
+ * address for a "connected" DGRAM socket. This is true for most, but
+ * probably not all systems. For some systems, this could be done
+ * cleaner by using the IP_RECVDSTADDR option + recvmsg().
+ */
+static int
+get_local_addr(struct sockaddr *remote, int remlen,
+	       struct sockaddr *local, socklen_t *loclen)
+{
+	int s, ret;
```
Comment (elric1) [inline kpasswd/kpasswdd.c]: Inconsistent indentation with the rest of the file.  This is NetBSD KNF which, e.g. has 8 space tabs.

File: lib/kadm5/ad.c (hunk 3 above)
Comment (jaltman) [inline lib/kadm5/ad.c]: Not necessary but no harm.

File: kpasswd/kpasswdd.c [hunk 5]
```diff
@@ -691,48 +717,67 @@ doit(krb5_keytab keytab, int port)
...
-	    krb5_warn(context, save_errno, "bind(%s)", str);
-	    continue;
-	}
-	maxfd = max(maxfd, sockets[i]);
-	if (maxfd >= FD_SETSIZE)
-	    krb5_errx(context, 1, "fd too large");
-	FD_SET(sockets[i], &real_fdset);
+        n = 1;
+        maxfd = 0;
+	fdz = 0;
```
Comment (elric1) [inline kpasswd/kpasswdd.c]: The fdz line has a tab.  Other lines have a space.

Comment (elric1) [inline kpasswd/kpasswdd.c]: I'm not sure that this is actually portable.  We are using stdin assuming that it works as input and output later in the program.
I'm also not comfortable that we don't dup2(2) it up to being larger than two as well as opening /dev/null and dup2(2) the resulting fd to 0, 1, and 2.

File: lib/hcrypto/libtommath/mtest/mpi.c [hunk 6]
```diff
@@ -492,7 +492,7 @@ void   mp_set(mp_int *mp, mp_digit d)
 mp_err mp_set_int(mp_int *mp, long z)
 {
   int            ix;
-  unsigned long  v = labs(z);
```
Comment (jaltman) [inline lib/hcrypto/libtommath/mtest/mpi.c]: This should remain labs() because 'z' is a long.

File: kpasswd/kpasswdd.c (hunk 5 above)
Comment (nicowilliams) [inline kpasswd/kpasswdd.c]: It is portable, but I agree that we should dup2() fildes zero out of the way and replace 0, 1, and 2 with /dev/null.

File: lib/kadm5/ad.c (hunk 2 above)
Comment (nicowilliams) [inline lib/kadm5/ad.c]: Use &w=1.

Comment (nicowilliams) [inline lib/kadm5/ad.c]: Or ?w=1

File: kpasswd/kpasswdd.c (hunk 5 above)
Comment (nicowilliams) [inline kpasswd/kpasswdd.c]: Fix pushed (but I'm only building now).

File: kpasswd/kpasswdd.c [hunk 7]
```diff
@@ -691,91 +717,144 @@ doit(krb5_keytab keytab, int port)
...
+        if (devnull != STDOUT_FILENO)
+            (void) dup2(devnull, STDOUT_FILENO);
+        if (devnull != STDERR_FILENO)
+            (void) dup2(devnull, STDERR_FILENO);
+        /* Now dup(stdin) -- must return something larger than stder */
+        if ((fdz = dup(STDIN_FILENO)) == -1)
+            krb5_errx(context, 1, "Could not dup(STDIN_FILENO)");
+        /* Replace stdin with /dev/null */
+        (void) dup2(devnull, STDIN_FILENO);
+        (void) close(devnull);
```
Comment (elric1) [inline kpasswd/kpasswdd.c]: if (devnull > 2)
    close(devnull);

Comment (nicowilliams) [inline kpasswd/kpasswdd.c]: Good catch.

File: kpasswd/kpasswdd.c [hunk 8]
```diff
@@ -646,6 +646,32 @@ process (krb5_keytab keytab,
...
+ * XXX this code relies on getsockname() returning a valid local
+ * address for a "connected" DGRAM socket. This is true for most, but
+ * probably not all systems. For some systems, this could be done
+ * cleaner by using the IP_RECVDSTADDR option + recvmsg().
+ */
+static int
+get_local_addr(struct sockaddr *remote, int remlen,
+	       struct sockaddr *local, socklen_t *loclen)
+{
+    int s, ret;
```
Comment (vdukhovni) [inline kpasswd/kpasswdd.c]: Probably instead `krb5_socket_t s` and if needed `#include <krb5-types.h>`

Comment (jaltman) [inline kpasswd/kpasswdd.c]: why krb5_socket_t?  normally for Heimdal we would use rk_socket_t, rk_INVALID_SOCKET, and rk_closesocket in cases where we want portability.

Comment (vdukhovni) [inline kpasswd/kpasswdd.c]: Those are fine too.  I would expect these to be the same, no?  But you're probably right.

File: kpasswd/kpasswdd.c [hunk 9]
```diff
@@ -691,91 +717,145 @@ doit(krb5_keytab keytab, int port)
     int *sockets;
```
Comment (vdukhovni) [inline kpasswd/kpasswdd.c]: Here also I guess, rk_socket_t or similar

**Comment by vdukhovni** [comment]
May be best to defer the `Add inetd support to kpasswdd` commit for now, and do it separately.  Or is all that code Unix-only?

**Comment by vdukhovni** [comment]
Otherwise, LGTM.

**Comment by jaltman** [comment]
kpasswdd  is not built on Windows at present.
